File: packages/nerva-js/src/context.ts
/**
 * ExecContext — the connective tissue that flows through every Nerva primitive.
 *
 * Every operation in Nerva receives an ExecContext. It carries identity, permissions,
 * observability (spans/events), token accounting, cancellation, and streaming state.
 * This is primitive #0: the context object that all other primitives depend on.
 *
 * @module context
 */

import { randomUUID } from "node:crypto";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Default scope for memory isolation when none is specified. */
const DEFAULT_MEMORY_SCOPE: Scope = "session";

// ---------------------------------------------------------------------------
// Scope (N-201)
// ---------------------------------------------------------------------------

/**
 * Memory isolation boundary for context data.
 *
 * Determines how far stored facts and state are visible:
 * - `user` — persists across sessions for the same user
 * - `session` — scoped to a single conversation session
 * - `agent` — private to the agent handling the request
 * - `global` — visible to all users and agents
 */
export type Scope = "user" | "session" | "agent" | "global";

/** All valid scope values for runtime validation. */
export const SCOPE_VALUES: readonly Scope[] = [
  "user",
  "session",
  "agent",
  "global",
] as const;

// ---------------------------------------------------------------------------
// Permissions (N-201)
// ---------------------------------------------------------------------------

/**
 * Immutable capability set governing what a context is allowed to do.
 *
 * Uses allowlists for tools and agents. A value of `null` means
 * "no restriction" (all allowed). An empty `ReadonlySet` means "none allowed".
 */
export interface Permissions {
  /** Role names assigned to this context (e.g. `"admin"`, `"user"`). */
  readonly roles: ReadonlySet<string>;

  /** Tool names this context may invoke, or `null` for unrestricted. */
  readonly allowedTools: ReadonlySet<string> | null;

  /** Agent names this context may delegate to, or `null` for unrestricted. */
  readonly allowedAgents: ReadonlySet<string> | null;

  /**
   * Check whether the given tool is permitted.
   *
   * @param toolName - Fully-qualified tool name to check.
   * @returns `true` if the tool is allowed (or if no restriction is set).
   */
  canUseTool(toolName: string): boolean;

  /**
   * Check whether delegation to the given agent is permitted.
   *
   * @param agentName - Agent identifier to check.
   * @returns `true` if the agent is allowed (or if no restriction is set).
   */
  canUseAgent(agentName: string): boolean;

  /**
   * Check whether the context carries a specific role.
   *
   * @param role - Role name to look for.
   * @returns `true` if the role is present.
   */
  hasRole(role: string): boolean;
}

/** Options for creating a {@link Permissions} instance. */
export interface PermissionsInit {
  readonly roles?: ReadonlySet<string>;
  readonly allowedTools?: ReadonlySet<string> | null;
  readonly allowedAgents?: ReadonlySet<string> | null;
}

/**
 * Create an immutable {@link Permissions} object.
 *
 * @param init - Optional initial values. Defaults to unrestricted.
 * @returns A frozen `Permissions` instance.
 */
export function createPermissions(init?: PermissionsInit): Permissions {
  const roles: ReadonlySet<string> = init?.roles ?? new Set<string>();
  const allowedTools: ReadonlySet<string> | null =
    init?.allowedTools ?? null;
  const allowedAgents: ReadonlySet<string> | null =
    init?.allowedAgents ?? null;

  return Object.freeze({
    roles,
    allowedTools,
    allowedAgents,

    canUseTool(toolName: string): boolean {
      if (allowedTools === null) return true;
      return allowedTools.has(toolName);
    },

    canUseAgent(agentName: string): boolean {
      if (allowedAgents === null) return true;
      return allowedAgents.has(agentName);
    },

    hasRole(role: string): boolean {
      return roles.has(role);
    },
  });
}

// ---------------------------------------------------------------------------
// TokenUsage (N-201)
// ---------------------------------------------------------------------------

/**
 * Immutable accumulator for LLM token consumption and estimated cost.
 *
 * Instances are never mutated — {@link TokenUsage.add} returns a new instance.
 */
export class TokenUsage {
  /** Number of tokens in the prompt/input. */
  readonly promptTokens: number;

  /** Number of tokens in the completion/output. */
  readonly completionTokens: number;

  /** Combined prompt + completion tokens. */
  readonly totalTokens: number;

  /** Estimated cost in US dollars. */
  readonly costUsd: number;

  /**
   * @param promptTokens - Prompt token count. Defaults to 0.
   * @param completionTokens - Completion token count. Defaults to 0.
   * @param totalTokens - Total token count. Defaults to 0.
   * @param costUsd - Estimated cost in USD. Defaults to 0.
   */
  constructor(
    promptTokens = 0,
    completionTokens = 0,
    totalTokens = 0,
    costUsd = 0,
  ) {
    this.promptTokens = promptTokens;
    this.completionTokens = completionTokens;
    this.totalTokens = totalTokens;
    this.costUsd = costUsd;
  }

  /**
   * Return a new `TokenUsage` that is the sum of this and `other`.
   * Neither operand is mutated.
   *
   * @param other - Token usage to add.
   * @returns A fresh `TokenUsage` with summed fields.
   */
  add(other: TokenUsage): TokenUsage {
    return new TokenUsage(
      this.promptTokens + other.promptTokens,
      this.completionTokens + other.completionTokens,
      this.totalTokens + other.totalTokens,
      this.costUsd + other.costUsd,
    );
  }
}

// ---------------------------------------------------------------------------
// Span & Event (N-201)
// ---------------------------------------------------------------------------

/**
 * A timed segment of work within a request's lifecycle.
 *
 * Spans form a tree: each span may have a `parentId` pointing to the
 * span that created it. A `null` parent means a root span.
 */
export interface Span {
  /** Unique identifier for this span. */
  readonly spanId: string;

  /** Human-readable label (e.g. `"llm.call"`, `"tool.invoke"`). */
  readonly name: string;

  /** Span ID of the parent, or `null` for root spans. */
  readonly parentId: string | null;

  /** Unix timestamp (seconds) when the span started. */
  readonly startedAt: number;

  /** Unix timestamp (seconds) when the span ended, or `null` if still open. */
  readonly endedAt: number | null;

  /** Arbitrary key-value metadata attached to the span. */
  readonly attributes: Readonly<Record<string, string>>;
}

/**
 * A point-in-time occurrence recorded within a context.
 *
 * Events are simpler than spans — they have no duration, just a timestamp
 * and descriptive metadata.
 */
export interface Event {
  /** Unix timestamp (seconds) of the event. */
  readonly timestamp: number;

  /** Human-readable label (e.g. `"policy.denied"`, `"stream.started"`). */
  readonly name: string;

  /** Arbitrary key-value metadata. */
  readonly attributes: Readonly<Record<string, string>>;
}

// ---------------------------------------------------------------------------
// StreamSink (N-202)
// ---------------------------------------------------------------------------

/**
 * Protocol for pushing incremental output chunks to a consumer.
 *
 * Implementations may write to an HTTP response, a WebSocket, a queue,
 * or an in-memory buffer (for testing).
 */
export interface StreamSink {
  /**
   * Send a single chunk of output.
   *
   * @param chunk - Text fragment to push downstream.
   */
  push(chunk: string): Promise<void>;

  /** Signal that no more chunks will be sent. */
  close(): Promise<void>;
}

/**
 * In-memory {@link StreamSink} implementation for testing.
 *
 * Collects all pushed chunks into a list so tests can assert
 * on the full output without I/O.
 */
export class InMemoryStreamSink implements StreamSink {
  /** All chunks pushed so far, in order. */
  readonly chunks: string[] = [];

  /** Whether {@link close} has been called. */
  private _closed = false;

  /** Whether the sink has been closed. */
  get closed(): boolean {
    return this._closed;
  }

  /**
   * Append a chunk to the internal buffer.
   *
   * @param chunk - Text fragment to record.
   * @throws {Error} If the sink has already been closed.
   */
  async push(chunk: string): Promise<void> {
    if (this._closed) {
      throw new Error("Cannot push to a closed StreamSink");
    }
    this.chunks.push(chunk);
  }

  /**
   * Mark the sink as closed. Subsequent pushes will throw.
   *
   * @throws {Error} If the sink has already been closed.
   */
  async close(): Promise<void> {
    if (this._closed) {
      throw new Error("StreamSink is already closed");
    }
    this._closed = true;
  }
}

// ---------------------------------------------------------------------------
// ExecContext (N-202)
// ---------------------------------------------------------------------------

/** Options for {@link ExecContext.create}. */
export interface ExecContextCreateOptions {
  /** Authenticated user identifier, or `null` for anonymous. */
  readonly userId?: string | null;

  /** Conversation/session identifier, or `null`. */
  readonly sessionId?: string | null;

  /** Capability set. Defaults to an unrestricted `Permissions`. */
  readonly permissions?: Permissions;

  /** Memory isolation boundary. Defaults to `"session"`. */
  readonly memoryScope?: Scope;

  /** Seconds from now until the context times out, or `null`. */
  readonly timeoutSeconds?: number | null;

  /** Optional sink for incremental output. */
  readonly stream?: StreamSink | null;
}

/**
 * Execution context that flows through every Nerva primitive.
 *
 * Carries identity, permissions, observability (spans and events),
 * token accounting, cancellation signalling, and an optional stream sink.
 * Contexts are created via the static {@link ExecContext.create} factory
 * and can spawn children for sub-operations via {@link ExecContext.child}.
 */
export class ExecContext {
  /** Unique identifier for this individual request. */
  readonly requestId: string;

  /** Groups related requests into a single trace. */
  readonly traceId: string;

  /** Authenticated user, or `null` for anonymous. */
  readonly userId: string | null;

  /** Conversation/session identifier, or `null`. */
  readonly sessionId: string | null;

  /** Capability set governing tool/agent access. */
  readonly permissions: Permissions;

  /** Isolation boundary for stored state. */
  readonly memoryScope: Scope;

  /** Ordered list of timed work segments. */
  readonly spans: Span[];

  /** Ordered list of point-in-time occurrences. */
  readonly events: Event[];

  /** Accumulated LLM token consumption. */
  tokenUsage: TokenUsage;

  /** Unix timestamp when this context was created. */
  readonly createdAt: number;

  /** Unix timestamp after which the context is timed out, or `null`. */
  readonly timeoutAt: number | null;

  /** Resolves when cancellation is signalled. */
  private readonly _cancelController: AbortController;

  /** Optional sink for incremental output. */
  stream: StreamSink | null;

  /** Arbitrary string tags for policy conditions and routing. */
  readonly metadata: Record<string, string>;

  /** Current delegation depth (0 = root, incremented by `child()`). */
  readonly depth: number;

  private constructor(params: {
    requestId: string;
    traceId: string;
    userId: string | null;
    sessionId: string | null;
    permissions: Permissions;
    memoryScope: Scope;
    spans: Span[];
    events: Event[];
    tokenUsage: TokenUsage;
    createdAt: number;
    timeoutAt: number | null;
    cancelController: AbortController;
    stream: StreamSink | null;
    metadata: Record<string, string>;
    depth: number;
  }) {
    this.requestId = params.requestId;
    this.traceId = params.traceId;
    this.userId = params.userId;
    this.sessionId = params.sessionId;
    this.permissions = params.permissions;
    this.memoryScope = params.memoryScope;
    this.spans = params.spans;
    this.events = params.events;
    this.tokenUsage = params.tokenUsage;
    this.createdAt = params.createdAt;
    this.timeoutAt = params.timeoutAt;
    this._cancelController = params.cancelController;
    this.stream = params.stream;
    this.metadata = params.metadata;
    this.depth = params.depth;
  }

  /**
   * Create a new root execution context.
   *
   * Generates fresh `requestId` and `traceId`, sets the clock,
   * and initialises all accumulators to empty.
   *
   * @param options - Optional creation parameters.
   * @returns A fully initialised `ExecContext` ready for use.
   */
  static create(options?: ExecContextCreateOptions): ExecContext {
    const now = Date.now() / 1000;
    const timeoutSeconds = options?.timeoutSeconds ?? null;
    const timeoutAt =
      timeoutSeconds !== null ? now + timeoutSeconds : null;

    return new ExecContext({
      requestId: randomUUID().replace(/-/g, ""),
      traceId: randomUUID().replace(/-/g, ""),
      userId: options?.userId ?? null,
      sessionId: options?.sessionId ?? null,
      permissions: options?.permissions ?? createPermissions(),
      memoryScope: options?.memoryScope ?? DEFAULT_MEMORY_SCOPE,
      spans: [],
      events: [],
      tokenUsage: new TokenUsage(),
      createdAt: now,
      timeoutAt,
      cancelController: new AbortController(),
      stream: options?.stream ?? null,
      metadata: {},
      depth: 0,
    });
  }

  /**
   * Create a child context for delegation to a sub-handler.
   *
   * The child inherits the parent's trace, permissions, memory scope,
   * timeout, cancellation signal, and stream — but gets a fresh
   * `requestId` and a new root span named after `handlerName`.
   *
   * @param handlerName - Label for the child operation (used as the span name).
   * @returns A new `ExecContext` linked to the same trace as the parent.
   */
  child(handlerName: string): ExecContext {
    const now = Date.now() / 1000;
    const rootSpan: Span = {
      spanId: randomUUID().replace(/-/g, ""),
      name: handlerName,
      parentId: this.requestId,
      startedAt: now,
      endedAt: null,
      attributes: {},
    };

    return new ExecContext({
      requestId: randomUUID().replace(/-/g, ""),
      traceId: this.traceId,
      userId: this.userId,
      sessionId: this.sessionId,
      permissions: this.permissions,
      memoryScope: this.memoryScope,
      spans: [rootSpan],
      events: [],
      tokenUsage: new TokenUsage(),
      createdAt: now,
      timeoutAt: this.timeoutAt,
      cancelController: this._cancelController,
      stream: this.stream,
      metadata: { ...this.metadata },
      depth: this.depth + 1,
    });
  }

  // -- Cancellation -------------------------------------------------------

  /**
   * The `AbortSignal` that fires when this context is cancelled.
   *
   * @returns The signal tied to this context's cancel controller.
   */
  get cancelSignal(): AbortSignal {
    return this._cancelController.signal;
  }

  /**
   * Signal cancellation for this context (and any children sharing
   * the same controller).
   */
  cancel(): void {
    this._cancelController.abort();
  }

  // -- Query helpers ------------------------------------------------------

  /**
   * Check whether the context has exceeded its timeout.
   *
   * @returns `true` if a timeout was set and the current time is past it.
   */
  isTimedOut(): boolean {
    if (this.timeoutAt === null) return false;
    return Date.now() / 1000 > this.timeoutAt;
  }

  /**
   * Check whether cancellation has been signalled.
   *
   * @returns `true` if the cancel controller has been aborted.
   */
  isCancelled(): boolean {
    return this._cancelController.signal.aborted;
  }

  /**
   * Seconds elapsed since this context was created.
   *
   * @returns Wall-clock seconds since `createdAt`.
   */
  elapsedSeconds(): number {
    return Date.now() / 1000 - this.createdAt;
  }

  // -- Mutation helpers (append-only) -------------------------------------

  /**
   * Start a new span and append it to this context's span list.
   *
   * The span is created with `endedAt = null` (still open). Callers
   * are responsible for closing it when the work completes.
   *
   * @param name - Human-readable label for the span.
   * @returns The newly created `Span`.
   */
  addSpan(name: string): Span {
    const span: Span = {
      spanId: randomUUID().replace(/-/g, ""),
      name,
      parentId: this.requestId,
      startedAt: Date.now() / 1000,
      endedAt: null,
      attributes: {},
    };
    this.spans.push(span);
    return span;
  }

  /**
   * Record a point-in-time event in this context.
   *
   * @param name - Human-readable label for the event.
   * @param attributes - Arbitrary string key-value pairs.
   * @returns The newly created `Event`.
   */
  addEvent(name: string, attributes: Record<string, string> = {}): Event {
    const event: Event = {
      timestamp: Date.now() / 1000,
      name,
      attributes: { ...attributes },
    };
    this.events.push(event);
    return event;
  }

  /**
   * Accumulate token usage into this context's running total.
   *
   * @param usage - Token counts and cost to add.
   */
  recordTokens(usage: TokenUsage): void {
    this.tokenUsage = this.tokenUsage.add(usage);
  }
}

File: packages/nerva-py/nerva/context.py
"""ExecContext — the connective tissue that flows through every Nerva primitive.

Every operation in Nerva receives an ExecContext. It carries identity, permissions,
observability (spans/events), token accounting, cancellation, and streaming state.
This is primitive #0: the context object that all other primitives depend on.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Protocol
from uuid import uuid4


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

DEFAULT_MEMORY_SCOPE = "session"
"""Default scope for memory isolation when none is specified."""


# ---------------------------------------------------------------------------
# Scope (N-102)
# ---------------------------------------------------------------------------


class Scope(StrEnum):
    """Memory isolation boundary for context data.

    Determines how far stored facts and state are visible:
    - USER: persists across sessions for the same user
    - SESSION: scoped to a single conversation session
    - AGENT: private to the agent handling the request
    - GLOBAL: visible to all users and agents
    """

    USER = "user"
    SESSION = "session"
    AGENT = "agent"
    GLOBAL = "global"


# ---------------------------------------------------------------------------
# Permissions (N-102)
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Permissions:
    """Immutable capability set governing what a context is allowed to do.

    Uses allowlists for tools and agents. A value of ``None`` means
    "no restriction" (all allowed). An empty frozenset means "none allowed".

    Attributes:
        roles: Set of role names assigned to this context (e.g. ``{"admin", "user"}``).
        allowed_tools: Tool names this context may invoke, or ``None`` for unrestricted.
        allowed_agents: Agent names this context may delegate to, or ``None`` for unrestricted.
    """

    roles: frozenset[str] = field(default_factory=frozenset)
    allowed_tools: frozenset[str] | None = None
    allowed_agents: frozenset[str] | None = None

    def can_use_tool(self, tool_name: str) -> bool:
        """Check whether the given tool is permitted.

        Args:
            tool_name: Fully-qualified tool name to check.

        Returns:
            ``True`` if the tool is allowed (or if no restriction is set).
        """
        if self.allowed_tools is None:
            return True
        return tool_name in self.allowed_tools

    def can_use_agent(self, agent_name: str) -> bool:
        """Check whether delegation to the given agent is permitted.

        Args:
            agent_name: Agent identifier to check.

        Returns:
            ``True`` if the agent is allowed (or if no restriction is set).
        """
        if self.allowed_agents is None:
            return True
        return agent_name in self.allowed_agents

    def has_role(self, role: str) -> bool:
        """Check whether the context carries a specific role.

        Args:
            role: Role name to look for.

        Returns:
            ``True`` if the role is present.
        """
        return role in self.roles


# ---------------------------------------------------------------------------
# TokenUsage (N-102)
# ---------------------------------------------------------------------------


@dataclass
class TokenUsage:
    """Accumulator for LLM token consumption and estimated cost.

    Attributes:
        prompt_tokens: Number of tokens in the prompt/input.
        completion_tokens: Number of tokens in the completion/output.
        total_tokens: Combined prompt + completion tokens.
        cost_usd: Estimated cost in US dollars.
    """

    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    cost_usd: float = 0.0

    def add(self, other: TokenUsage) -> TokenUsage:
        """Return a new ``TokenUsage`` that is the sum of *self* and *other*.

        Neither operand is mutated.

        Args:
            other: Token usage to add.

        Returns:
            A fresh ``TokenUsage`` with summed fields.
        """
        return TokenUsage(
            prompt_tokens=self.prompt_tokens + other.prompt_tokens,
            completion_tokens=self.completion_tokens + other.completion_tokens,
            total_tokens=self.total_tokens + other.total_tokens,
            cost_usd=self.cost_usd + other.cost_usd,
        )


# ---------------------------------------------------------------------------
# Span & Event (N-102)
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Span:
    """A timed segment of work within a request's lifecycle.

    Spans form a tree: each span may have a ``parent_id`` pointing to
    the span that created it. A ``None`` parent means a root span.

    Attributes:
        span_id: Unique identifier for this span.
        name: Human-readable label (e.g. ``"llm.call"``, ``"tool.invoke"``).
        parent_id: Span ID of the parent, or ``None`` for root spans.
        started_at: Unix timestamp when the span started.
        ended_at: Unix timestamp when the span ended, or ``None`` if still open.
        attributes: Arbitrary key-value metadata attached to the span.
    """

    span_id: str
    name: str
    parent_id: str | None
    started_at: float
    ended_at: float | None
    attributes: dict[str, str]


@dataclass(frozen=True)
class Event:
    """A point-in-time occurrence recorded within a context.

    Events are simpler than spans — they have no duration, just a timestamp
    and descriptive metadata.

    Attributes:
        timestamp: Unix timestamp of the event.
        name: Human-readable label (e.g. ``"policy.denied"``, ``"stream.started"``).
        attributes: Arbitrary key-value metadata.
    """

    timestamp: float
    name: str
    attributes: dict[str, str]


# ---------------------------------------------------------------------------
# StreamSink (N-103)
# ---------------------------------------------------------------------------


class StreamSink(Protocol):
    """Protocol for pushing incremental output chunks to a consumer.

    Implementations may write to an HTTP response, a WebSocket, a queue,
    or an in-memory buffer (for testing).
    """

    async def push(self, chunk: str) -> None:
        """Send a single chunk of output.

        Args:
            chunk: Text fragment to push downstream.
        """
        ...

    async def close(self) -> None:
        """Signal that no more chunks will be sent."""
        ...


class InMemoryStreamSink:
    """In-memory ``StreamSink`` implementation for testing.

    Collects all pushed chunks into a list so tests can assert
    on the full output without I/O.

    Attributes:
        chunks: All chunks pushed so far, in order.
        closed: Whether ``close()`` has been called.
    """

    def __init__(self) -> None:
        self.chunks: list[str] = []
        self.closed: bool = False

    async def push(self, chunk: str) -> None:
        """Append *chunk* to the internal buffer.

        Args:
            chunk: Text fragment to record.

        Raises:
            RuntimeError: If the sink has already been closed.
        """
        if self.closed:
            raise RuntimeError("Cannot push to a closed StreamSink")
        self.chunks.append(chunk)

    async def close(self) -> None:
        """Mark the sink as closed. Subsequent pushes will raise.

        Raises:
            RuntimeError: If the sink has already been closed.
        """
        if self.closed:
            raise RuntimeError("StreamSink is already closed")
        self.closed = True


# ---------------------------------------------------------------------------
# ExecContext (N-101)
# ---------------------------------------------------------------------------


@dataclass
class ExecContext:
    """Execution context that flows through every Nerva primitive.

    Carries identity, permissions, observability (spans and events),
    token accounting, cancellation signalling, and an optional stream sink.
    Contexts are created via the ``create`` factory and can spawn children
    for sub-operations via ``child``.

    Attributes:
        request_id: Unique identifier for this individual request.
        trace_id: Groups related requests into a single trace.
        user_id: Authenticated user, or ``None`` for anonymous.
        session_id: Conversation/session identifier, or ``None``.
        permissions: Capability set governing tool/agent access.
        memory_scope: Isolation boundary for stored state.
        spans: Ordered list of timed work segments.
        events: Ordered list of point-in-time occurrences.
        token_usage: Accumulated LLM token consumption.
        created_at: Unix timestamp when this context was created.
        timeout_at: Unix timestamp after which the context is timed out, or ``None``.
        cancelled: Async event that signals cancellation when set.
        stream: Optional sink for incremental output.
        metadata: Arbitrary string tags for policy conditions and routing.
    """

    request_id: str
    trace_id: str
    user_id: str | None
    session_id: str | None
    permissions: Permissions
    memory_scope: Scope
    spans: list[Span]
    events: list[Event]
    token_usage: TokenUsage
    created_at: float
    timeout_at: float | None
    cancelled: asyncio.Event
    stream: StreamSink | None
    metadata: dict[str, str]
    depth: int = 0

    @classmethod
    def create(
        cls,
        user_id: str | None = None,
        session_id: str | None = None,
        permissions: Permissions | None = None,
        memory_scope: Scope = Scope.SESSION,
        timeout_seconds: float | None = None,
        stream: StreamSink | None = None,
    ) -> ExecContext:
        """Create a new root execution context.

        Generates fresh ``request_id`` and ``trace_id``, sets the clock,
        and initialises all accumulators to empty.

        Args:
            user_id: Authenticated user identifier, or ``None`` for anonymous.
            session_id: Conversation/session identifier, or ``None``.
            permissions: Capability set. Defaults to an unrestricted ``Permissions``.
            memory_scope: Memory isolation boundary. Defaults to ``Scope.SESSION``.
            timeout_seconds: Seconds from now until the context times out, or ``None``.
            stream: Optional sink for incremental output.

        Returns:
            A fully initialised ``ExecContext`` ready for use.
        """
        now = time.time()
        timeout_at = (now + timeout_seconds) if timeout_seconds is not None else None

        return cls(
            request_id=uuid4().hex,
            trace_id=uuid4().hex,
            user_id=user_id,
            session_id=session_id,
            permissions=permissions or Permissions(),
            memory_scope=memory_scope,
            spans=[],
            events=[],
            token_usage=TokenUsage(),
            created_at=now,
            timeout_at=timeout_at,
            cancelled=asyncio.Event(),
            stream=stream,
            metadata={},
        )

    def child(self, handler_name: str) -> ExecContext:
        """Create a child context for delegation to a sub-handler.

        The child inherits the parent's trace, permissions, memory scope,
        timeout, cancellation signal, and stream — but gets a fresh
        ``request_id``, a new root span named after *handler_name*,
        and a depth incremented by one.

        Args:
            handler_name: Label for the child operation (used as the span name).

        Returns:
            A new ``ExecContext`` linked to the same trace as the parent.
        """
        child_request_id = uuid4().hex
        root_span = Span(
            span_id=uuid4().hex,
            name=handler_name,
            parent_id=self.request_id,
            started_at=time.time(),
            ended_at=None,
            attributes={},
        )

        return ExecContext(
            request_id=child_request_id,
            trace_id=self.trace_id,
            user_id=self.user_id,
            session_id=self.session_id,
            permissions=self.permissions,
            memory_scope=self.memory_scope,
            spans=[root_span],
            events=[],
            token_usage=TokenUsage(),
            created_at=time.time(),
            timeout_at=self.timeout_at,
            cancelled=self.cancelled,
            stream=self.stream,
            metadata=dict(self.metadata),
            depth=self.depth + 1,
        )

    # -- Query helpers -----------------------------------------------------

    def is_timed_out(self) -> bool:
        """Check whether the context has exceeded its timeout.

        Returns:
            ``True`` if a timeout was set and the current time is past it.
        """
        if self.timeout_at is None:
            return False
        return time.time() > self.timeout_at

    def is_cancelled(self) -> bool:
        """Check whether cancellation has been signalled.

        Returns:
            ``True`` if the ``cancelled`` event has been set.
        """
        return self.cancelled.is_set()

    def elapsed_seconds(self) -> float:
        """Seconds elapsed since this context was created.

        Returns:
            Wall-clock seconds since ``created_at``.
        """
        return time.time() - self.created_at

    # -- Mutation helpers (append-only) ------------------------------------

    def add_span(self, name: str) -> Span:
        """Start a new span and append it to this context's span list.

        The span is created with ``ended_at=None`` (still open). Callers
        are responsible for closing it when the work completes.

        Args:
            name: Human-readable label for the span.

        Returns:
            The newly created ``Span``.
        """
        span = Span(
            span_id=uuid4().hex,
            name=name,
            parent_id=self.request_id,
            started_at=time.time(),
            ended_at=None,
            attributes={},
        )
        self.spans.append(span)
        return span

    def add_event(self, name: str, **attributes: str) -> Event:
        """Record a point-in-time event in this context.

        Args:
            name: Human-readable label for the event.
            **attributes: Arbitrary string key-value pairs attached to the event.

        Returns:
            The newly created ``Event``.
        """
        event = Event(
            timestamp=time.time(),
            name=name,
            attributes=dict(attributes),
        )
        self.events.append(event)
        return event

    def record_tokens(self, usage: TokenUsage) -> None:
        """Accumulate token usage into this context's running total.

        Args:
            usage: Token counts and cost to add.
        """
        self.token_usage = self.token_usage.add(usage)
